src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -854,8 +857,15 @@ private static JsonSerializerOptions GetOrCreateDefaultOptionsInstance()
         {
             var options = new JsonSerializerOptions
             {
-                TypeInfoResolver = DefaultJsonTypeInfoResolver.RootDefaultInstance(),
-                _isReadOnly = true
+                // Because're marking the default instance as read-only,
+                // we need to specify a resolver instance for the case where
+                // reflection is disabled by default: use one that returns null for all types.
+
+                TypeInfoResolver = AppContextSwitchHelper.IsDefaultReflectionDisabled
+                    ? JsonTypeInfoResolver.Combine()
+                    : DefaultJsonTypeInfoResolver.RootDefaultInstance(),
```

Comment:
We should add a trim test which validates that:
* If I set `IsDefaultReflectionDisabled` to true and use only the correct APIs I don't get a trim analysis warning.
* Setting the feature switch and using the correct APIs leads to the reflection codebase being trimmed away.

/cc @eerhardt as he has experience with writing these.

Reply:
Good point. It isn't entirely clear to me how we could test that it works as advertised from within dotnet/runtime.

Reply:
> Setting the feature switch and using the correct APIs leads to the reflection codebase being trimmed away.

I believe there is only 1 test like this so far: InvariantGlobalization.

You can set the feature switch like this:

https://github.com/dotnet/runtime/blob/d2111a9baa0797424f2d5d2631bb21815972fef4/src/libraries/System.Runtime/tests/TrimmingTests/System.Runtime.TrimmingTests.proj#L18-L20

And then in the TrimmingTest app, you have to be careful you don't tip off the trimmer and it forces to keep the type, but you can check that the things are trimmed like this:

https://github.com/dotnet/runtime/blob/96a003edd9e25a7b4b693b6a7717ad4d46937a8b/src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationTrue.cs#L31-L67

> If I set IsDefaultReflectionDisabled to true and use only the correct APIs I don't get a trim analysis warning.

Unfortunately our trimming tests don't ensure there are no warnings today (because we do stuff like the above, which obviously cause warnings). But we should be able to opt-in to this functionality easily in the infrastructure by setting `TreatWarningsAsErrors=true` in the trimming test's "EnabledProperties", like we do for `EventSourceSupport` here:

https://github.com/dotnet/runtime/blob/d2111a9baa0797424f2d5d2631bb21815972fef4/src/libraries/System.Diagnostics.DiagnosticSource/tests/NativeAotTests/System.Diagnostics.DiagnosticSource.NativeAotTests.proj#L5-L6

You should just be able to say: `EnabledProperties="TreatWarningsAsErrors"`.

Reply:
> I'm guessing that there's a unit test somewhere else that calls it and verifies that the status code gives the expected value?

Correct, that's handled by the trimming test infrastructure:

https://github.com/dotnet/runtime/blob/5b47cd752078ad2cbc59c422536772a843a95586/eng/testing/linker/trimmingTests.targets#L144-L148

Reply:
Should I just add a console app to `System.Runtime.TrimmingTests`? Or is it only scoped to just S.P.CorLib APIs?

Reply:
Since this is STJ tests, you should add them to:

https://github.com/dotnet/runtime/tree/main/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/TrimmingTests

`System.Runtime.TrimmingTests` is for APIs exposed by `System.Runtime.dll`.